visual_reasoning_nemo_collection/text/tokenizer.py

```diff
@@ -0,0 +1,56 @@
+# Copyright (C) IBM Corporation 2018
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+import torch
+from torch import nn
+from nemo.backends.pytorch import NonTrainableNM
+from nemo.core.neural_types import NeuralType, ChannelType, LengthsType
+from nemo.utils.decorators import add_port_docs
+from nemo.collections.nlp.data import WordTokenizer
+
+class Tokenizer(NonTrainableNM):
+    """
+    Wrapping `nemo.collections.nlp.data.WordTokenizer`:
+    - Tokenizing text directly to ids.
+    - Sorts the sentences in decreasing length order to make torch's `pack_padded_sequence` happy.
+    - Applies `torch.nn.utils.rnn.pad_sequence`.
+    - Outputs everything in tensor format.
+    """
+    @property
+    @add_port_docs()
+    def input_ports(self):
+        """Returns definitions of module input ports.
+        """
+        return {
+            "x": NeuralType(('B', 'T'), ChannelType()),
+        }
+
+    @property
+    @add_port_docs()
+    def output_ports(self):
+        """Returns definitions of module output ports.
+        """
+        return {
+            "output": NeuralType(('B', 'T'), ChannelType()),
+            "output_len": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    def __init__(self, vocab_path: str):
+        super().__init__()
+
+        assert os.path.isfile(vocab_path)
+        self._tokenizer = WordTokenizer(vocab_path)
+
+    def forward(self, x):
+        tokenized = [self._tokenizer.text_to_ids(elem) for elem in x]
+        output_lens = [len(elem) for elem in tokenized]
+        # Sort in decreasing length order to make pack_padded_sequence happy
```

Comment:
Hey, this seems to be an error...! (That I also made one day)
Do I get it right that you changing the order of the samples in batch...?

https://github.com/IBM/pytorchpipe/blob/9cb17271666061cb19fe24197ecd5e4c8d32c5da/ptp/components/language/sentence_indexer.py#L142

Reply:
Yup it's an error, won't work correctly.
However I'm still trying to find a solution to this:
```
Traceback (most recent call last):
  File "baseline_model.py", line 155, in <module>
    neural_factory.train(training_graph=training_graph, callbacks=[callback, ecallback], optimization_params={"num_epochs": 10, "lr": 0.001}, optimizer="adam"
  File "/home/alexis/nemo/NeMo/nemo/core/neural_factory.py", line 598, in train
    amp_max_loss_scale=amp_max_loss_scale,
  File "/home/alexis/nemo/NeMo/nemo/backends/pytorch/actions.py", line 1415, in train
    call_chain=curr_call_chain, registered_tensors=registered_tensors,
  File "/home/alexis/nemo/NeMo/nemo/backends/pytorch/actions.py", line 442, in __nm_graph_forward_pass
    new_tensors = pmodule(force_pt=True, **call_set)
  File "/home/alexis/nemo/NeMo/nemo/backends/pytorch/nm.py", line 52, in __call__
    return nn.Module.__call__(self, *input, **kwargs)
  File "/home/alexis/miniconda3/envs/nemo/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/alexis/visual-reasoning/visual_reasoning_nemo_collection/misc/rnn.py", line 270, in forward
    embedded = nn.utils.rnn.pack_padded_sequence(embedded, input_lens, batch_first=True)
  File "/home/alexis/miniconda3/envs/nemo/lib/python3.7/site-packages/torch/nn/utils/rnn.py", line 233, in pack_padded_sequence
    _VF._pack_padded_sequence(input, lengths, batch_first)
RuntimeError: `lengths` array must be sorted in decreasing order when `enforce_sorted` is True. You can pass `enforce_sorted=False` to pack_padded_sequence and/or pack_sequence to sidestep this requirement if you do not need ONNX exportability.
```